src/librustc_trait_selection/traits/project.rs

```diff
@@ -1258,6 +1265,42 @@ fn confirm_generator_candidate<'cx, 'tcx>(
         .with_addl_obligations(obligations)
 }
 
+fn confirm_discriminant_kind_candidate<'cx, 'tcx>(
+    selcx: &mut SelectionContext<'cx, 'tcx>,
+    obligation: &ProjectionTyObligation<'tcx>,
+    discriminant_kind_vtable: VtableDiscriminantKindData<'tcx, PredicateObligation<'tcx>>,
+) -> Progress<'tcx> {
+    let tcx = selcx.tcx();
+
+    let enum_ty = selcx.infcx().shallow_resolve(discriminant_kind_vtable.enum_ty);
```

Comment:
You can get this type from `obligation`, not sure we need to have a copy of it in the `VtableDiscriminantKindData`, but @nikomatsakis would know better.

Also, general nit: `enum_ty` is confusing since this is not limited to `enum`s, `self_ty` would be better IMO.